conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/OptimizerTimeTracker.scala

```diff
@@ -0,0 +1,126 @@
+
+  /**
+    * Returns a dataframe prepared to apply the Catalyst workaround
+    */
+  def getWorkaroundDataframe: DataFrame = {
+    if (isWorkaroundEnabled) {
+      dfWithId
+    } else {
+      inputDf
+    }
+  }
+
+  /**
+    * Cleans up the additional field used for applying the Catalyst workaround
+    */
+  def cleanupWorkaroundDf(df: DataFrame): DataFrame = df.drop(col(idField1))
+
+  /**
+    * Returns true of a dataframe might require a Catalyst issue workaround.
+    * A workaround is required if execution plan optimization step takes too long
+    *
+    * @param df A dataframe that might require a Catalyst workaround
+    * @return true if a workaround is required
+    */
+  def isCatalystWorkaroundRequired(df: DataFrame, rulesApplied: Int): Boolean = {
+    if (isWorkaroundEnabled) {
+      val elapsedTime = getExecutionPlanGenerationTimeMs(df)
+
+      val tooBigTimeDifference = if (lastExecutionPlanOptimizatioTime > 0) {
+        elapsedTime / lastExecutionPlanOptimizatioTime > 5
+      } else {
+        false
+      }
+      val tooLong = tooBigTimeDifference ||
+        elapsedTime > baselineTimeMs + maxToleratedPlanGenerationPerRuleMs * rulesApplied
+
+      val msg = s"Execution optimization time for $rulesApplied rules: $elapsedTime ms"
+      if (tooLong) {
+        log.warn(s"$msg (Too long!)")
+      } else {
+        baselineTimeMs = Math.max(baselineTimeMs, elapsedTime)
+        lastExecutionPlanOptimizatioTime = elapsedTime
+        log.info(s"New baseline execution plan optimization time: $baselineTimeMs ms")
+        log.warn(msg)
+      }
+      tooLong
+    } else {
+      false
+    }
+  }
+
+  /**
+    * Records execution plan optimization time.
+    * This is used after a workarund and a conformance rule are applied to a dataframe.
+    *
+    * @param df A dataframe for measuring execution plan optimization time
+    */
+  def recordExecutionPlanOptimizationTime(df: DataFrame): Unit = {
+    val elapsedTime = getExecutionPlanGenerationTimeMs(df)
+    baselineTimeMs = Math.max(baselineTimeMs, elapsedTime)
+    lastExecutionPlanOptimizatioTime = elapsedTime
+  }
+
+  /**
+    * Returns elapsed time of execution plan optimization
+    *
+    * @param df A dataframe to calculate execution plan optimization time
+    * @return Elapsed time in milliseconds
+    */
+  def getExecutionPlanGenerationTimeMs(df: DataFrame): Long = {
+    val t0 = System.currentTimeMillis()
+    df.queryExecution.toString()
+    val t1 = System.currentTimeMillis()
+    t1 - t0
+  }
+
+  /**
+    * Applies a Catalyst workaround by joining a dataframe with the dataframe containing only unique ids.
+    *
+    * @param dfWithId A dataframe containing a unique id field for which execution plan generation takes too long
+    * @return A new dataframe with Catalyst optimizer issue workaround applied
+    */
+  def applyCatalystWorkaround(dfWithId: DataFrame): DataFrame = {
+    if (isWorkaroundEnabled) {
```

Comment:
**Minor**: Considering the name of the function, the check seems to counter the purpose. (also in the natural flow of things, it's double-checking the same condition - first in checking if the workaround is needed and then before actually applying it)

Reply:
Agree, the logic is duplicated a little. But I'd like to keep it that way so that `isWorkaroundEnabled ` be a big sledgehammer that turn everything on or off even if checker's and applicator's logic is changed independently.